### PR TITLE
Support creating a DatasetPipeline windowed by bytes

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds the ability to create a pipeline windowed by bytes, which simplifies many user calculations compared to creating it by blocks.

## Related issue number

Closes https://github.com/ray-project/ray/issues/18100

### CODE REVIEW

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -2603,6 +2608,9 @@ def window(self, *, blocks_per_window: int = 10) -> "DatasetPipeline[T]":
                 increases the latency to initial output, since it decreases the
                 length of the pipeline. Setting this to infinity effectively
                 disables pipelining.
+            bytes_per_window: Specify the window size in bytes instead of blocks.
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/dataset.py:2611]: I think it's a little complex (and maybe not needed once we have block splitting).

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -2638,7 +2646,27 @@ def gen():
...
         class Iterable:
             def __init__(self, blocks, epoch):
-                self._splits = blocks.split(split_size=blocks_per_window)
+                if bytes_per_window:
+                    self._splits = blocks.split_by_bytes(bytes_per_window)
+                else:
+                    self._splits = blocks.split(split_size=blocks_per_window)
+                sizes = [s.size_bytes() for s in self._splits]
+
+                def fmt(size_bytes):
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/data/dataset.py]: Done. Currently it should always be known.

**Review by ericl — COMMENTED** [review]
Comments addressed.

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -2638,7 +2653,36 @@ def gen():
 
         class Iterable:
             def __init__(self, blocks, epoch):
-                self._splits = blocks.split(split_size=blocks_per_window)
+                if bytes_per_window:
+                    self._splits = blocks.split_by_bytes(bytes_per_window)
+                else:
+                    self._splits = blocks.split(split_size=blocks_per_window)
+                try:
+                    sizes = [s.size_bytes() for s in self._splits]
```
Comment (jjyao) [inline python/ray/data/dataset.py:2661]: size_bytes() returns -1 if the size is unknown instead of exception? maybe just `assert min(sizes) >= 0`?

**Review by jjyao — APPROVED** [review]

**Pull request merged**